### PR TITLE
Synchronize access to unique_callstacks_ in SamplingProfiler (mostly)

### PR DESCRIPTION
**Pull request by dimitry-**
This map is modified from the capture thread while accessed from the
main thread which could cause undefined behavior. This commit
introduces a Mutex that guards access to unique_callstacks_ map
Note that this is not final solution it is rather temporary fix
until we move to more safe Capture Model.

### CODE REVIEW

File: OrbitGl/CaptureSerializer.cpp [hunk 1]
```diff
@@ -79,16 +79,16 @@ void CaptureSerializer::FillCaptureData(CaptureInfo* capture_info) {
...
-    CallstackInfo* callstack = capture_info->add_callstacks();
-    *callstack->mutable_data() = {hashed_callstack->m_Data.begin(),
-                                  hashed_callstack->m_Data.end()};
-  }
+  // TODO: this is not really synchronized, since GetCallstacks processing below
+  // is not under the same mutex lock we could end up having list of callstacks
+  // inconsistent with unique_callstacks. Revisit sampling profiler data
+  // thread-safety.
+  Capture::GSamplingProfiler->ForEachUniqueCallstack(
+      [&capture_info](const CallStack* call_stack) {
```
Comment (IrinaShkviro) [inline OrbitGl/CaptureSerializer.cpp:87]: Shouldn't be const CallStack& call_stack?

Comment (dimitry-) [reply OrbitGl/CaptureSerializer.cpp:87]: done.

File: OrbitCore/SamplingProfiler.h [hunk 2]
```diff
@@ -92,9 +94,12 @@ class SamplingProfiler {
     return &m_Callstacks;
   }
 
-  const std::unordered_map<CallstackID, std::shared_ptr<CallStack>>&
-  GetUniqueCallstacks() const {
-    return m_UniqueCallstacks;
+  template <typename F>
```
Comment (dpallotti) [inline OrbitCore/SamplingProfiler.h:97]: Why the template instead of passing a `std::function<void(const CallStack&)>`?

Comment (dimitry-) [reply OrbitCore/SamplingProfiler.h:97]: habbit from before std::function. fixed.

File: OrbitCore/SamplingProfiler.h [hunk 3]
```diff
@@ -92,9 +94,11 @@ class SamplingProfiler {
     return &m_Callstacks;
   }
 
-  const std::unordered_map<CallstackID, std::shared_ptr<CallStack>>&
-  GetUniqueCallstacks() const {
-    return m_UniqueCallstacks;
+  void ForEachUniqueCallstack(std::function<void(const CallStack&)> action) {
```
Comment (dpallotti) [inline OrbitCore/SamplingProfiler.h:97]: Same line as before, sorry... Can `action` be passed as const-ref?

Comment (dimitry-) [reply OrbitCore/SamplingProfiler.h]: done.

**Pull request merged**